### PR TITLE
Implement keep-alive feature.

### LINKED ISSUE #59: Getting io.grpc.StatusRuntimeException: UNAVAILABLE: Network closed for unknown reason
Hi,

I am creating a persistence subscription listener and am able to listen to the events produced, but the listeners are throwing the following error after an hour or so:

```
io.grpc.StatusRuntimeException: UNAVAILABLE: Network closed for unknown reason
```

Stacktrace:
```
exception io.grpc.StatusRuntimeException: UNAVAILABLE: Network closed for unknown re 
 ason                                                                                                                                                                                                                                                         
 eventstore     at io.grpc.Status.asRuntimeException(Status.java:533)                                                                                                                                                                                         
 eventstore     at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:478)                                                                                                                                                 
 eventstore     at io.grpc.internal.DelayedClientCall$DelayedListener$3.run(DelayedClientCall.java:464)                                                                                                                                                       
 eventstore     at io.grpc.internal.DelayedClientCall$DelayedListener.delayOrExecute(DelayedClientCall.java:428)                                                                                                                                              
 eventstore     at io.grpc.internal.DelayedClientCall$DelayedListener.onClose(DelayedClientCall.java:461)                                                                                                                                                     
 eventstore     at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:617)                                                                                                                                                                     
 eventstore     at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)                                                                                                                                                                         
 eventstore     at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:803)                                                                                                                                
 eventstore     at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:782)                                                                                                                               
 eventstore     at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)                                                                                                                                                                              
 eventstore     at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)                                                                                                                                                                     
 eventstore     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)                                                                                                                                                  
 eventstore     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)                                                                                                                                                  
 eventstore     at java.base/java.lang.Thread.run(Thread.java:834)
```

I'm using the following dependency:
```
        <dependency>
            <groupId>com.eventstore</groupId>
            <artifactId>db-client-java</artifactId>
            <version>0.5</version>
        </dependency>
```

### PR DESCRIPTION
**Pull request by YoEight**
Fixes: #59

### CODE REVIEW

File: db-client-java/src/main/java/com/eventstore/dbclient/ConnectionBuilder.java [hunk 1]
```diff
@@ -12,6 +12,7 @@
 public class ConnectionBuilder {
     private Timeouts _timeouts;
     private SslContext _sslContext = null;
+    private boolean _keepAlive = false;
```
Comment (oskardudycz) [inline db-client-java/src/main/java/com/eventstore/dbclient/ConnectionBuilder.java]: **QUESTION:** I'm not sure if that was the conclusion but didn't we say that we should have `keepAlive` set as true by default?

Comment (oskardudycz) [reply db-client-java/src/main/java/com/eventstore/dbclient/ConnectionBuilder.java]: Shouldn't we also allow the user to set `keepalive_time_ms` and `keepalive_timeout_ms`?

**Review by oskardudycz — CHANGES_REQUESTED** [review]
This needs alignment with the RFC: https://github.com/EventStore/home/blob/master/RFCs/016%20-%20Keep%20Alive.md.

**Comment by YoEight** [comment]
@oskardudycz the PR follows the specs now

File: db-client-java/src/main/java/com/eventstore/dbclient/ConnectionSettingsBuilder.java [hunk 2]
```diff
@@ -78,4 +81,14 @@ public ConnectionSettingsBuilder addHost(Endpoint host) {
         this._hosts.push(host);
         return this;
     }
+
+    public ConnectionSettingsBuilder keepAliveTimeout(long value) {
```
Comment (oskardudycz) [inline db-client-java/src/main/java/com/eventstore/dbclient/ConnectionSettingsBuilder.java:85]: **FIX:** Could we kee the naming consistent? In the builder above (https://github.com/EventStore/EventStoreDB-Client-Java/pull/60/files#diff-38d3d73124a50028533ed8eb19d8f761973f1aeafee7044034532e741fcf9ea9R32), we're using `keepAliveTimeoutInMs` while here we don't have `Ms` suffix. As other time-based settings are always provided with milliseconds, I'd suggest removing the `Ms` suffix.

The same suggestion applies to `keepAliveInterval`.

**Comment by oskardudycz** [comment]
@YoEight I don't see the requirements about the warning implemented:
> if the keepAliveInterval is smaller than ten seconds and bigger or equal 0 <0, 10), then we client or server should log warning information:
>
>$"Specified KeepAliveInterval of {keepAliveInterval} is less than recommended 10_000 ms"

Also, we're throwing ParseError with a generic error while RFC states:

> if the keepAliveInterval is smaller than -1 then error will be thrown/returned:
> 
> $"Invalid keepAliveInterval {keepAliveInterval}. Please provide a positive integer, or -1 to disable.
> 
> if the keepAliveTimeout is smaller than -1 then error will be thrown/returned:
>
> $"Invalid keepAliveTimeout {keepAliveTimeout}. Please provide a positive integer, or -1 to disable.

Other clients applied the above recommendations.

File: db-client-java/src/test/java/com/eventstore/dbclient/ParseValidConnectionStringTests.java [hunk 3]
```diff
@@ -91,7 +91,11 @@ public static Collection validConnectionStrings() {
                 {
                         "esdb://host?MaxDiscoverAttempts=200&discoveryinterval=1000&GOSSIPTIMEOUT=1&nOdEpReFeReNcE=leader&TLS=false&TlsVerifyCert=false&THROWOnAppendFailure=false",
                         "{\"dnsDiscover\":false,\"maxDiscoverAttempts\":200,\"discoveryInterval\":1000,\"gossipTimeout\":1,\"nodePreference\":\"leader\",\"tls\":false,\"tlsVerifyCert\":false,\"throwOnAppendFailure\":false,\"hosts\":[{\"address\":\"host\",\"port\":2113}]}"
-                }
+                },
+                {
+                        "esdb://localhost?keepAliveTimeout=20&keepAliveInterval=10",
```
Comment (oskardudycz) [inline db-client-java/src/test/java/com/eventstore/dbclient/ParseValidConnectionStringTests.java:96]: **CONCERN:** Can we add more test scenarios? E.g.
- when values equal to `-1` are provided,
- when values smaller than `-1` are provided,
- plus some permutations with valid and invalid values.

File: db-client-java/src/main/java/com/eventstore/dbclient/ConnectionSettingsBuilder.java (hunk 2 above)
Comment (YoEight) [reply db-client-java/src/main/java/com/eventstore/dbclient/ConnectionSettingsBuilder.java:85]: Done

**Comment by YoEight** [comment]
@oskardudycz I applied all your remarks, please take another look!

File: db-client-java/src/test/java/com/eventstore/dbclient/ParseValidConnectionStringTests.java (hunk 3 above)
Comment (YoEight) [reply db-client-java/src/test/java/com/eventstore/dbclient/ParseValidConnectionStringTests.java:96]: Done

File: db-client-java/src/main/java/com/eventstore/dbclient/EventStoreDBConnectionString.java [hunk 4]
```diff
@@ -263,8 +266,77 @@ private EventStoreDBClientSettings parseSearchParams(boolean first) throws Parse
                 this.settings.throwOnAppendFailure(value.equals("true"));
                 break;
             }
+            case "keepAliveTimeout": {
+                try {
+                    long parsedValue = parseLong(value);
```
Comment (oskardudycz) [inline db-client-java/src/main/java/com/eventstore/dbclient/EventStoreDBConnectionString.java:271]: **SUGGESTION:** It's not a blocker, might be handled eventually in the separate PR, but it might be worth to create a helper method for parsing those values to reduce the code redundancy. The same comment applies to the builder code.

**Review by oskardudycz — APPROVED** [review]

**Review by hayley-jean — APPROVED** [review]

File: db-client-java/src/main/java/com/eventstore/dbclient/EventStoreDBConnectionString.java (hunk 4 above)
Comment (YoEight) [reply db-client-java/src/main/java/com/eventstore/dbclient/EventStoreDBConnectionString.java:271]: Added it in the backlog: https://github.com/EventStore/EventStoreDB-Client-Java/issues/61
